### ISSUE TITLE
Example usage

### ISSUE THREAD

**Issue by mikesun** [networkprotocol/libyojimbo#25]
I'm trying to understand how to use the library, but have found the myself confused by the example client and server in `tests/`.

In the loop in `tests/client.cpp` (similar one in `tests/server.cpp`), the client establishes a connection to the server and continuously they then continuously exchange packets of type `TEST_PACKET_CONNECTION`. No user packets are exchanged however (type=`TEST_USER_PACKET`).

```
    while ( !quit )
    {
        client.SendPackets();

        clientTransport.WritePackets();

        clientTransport.ReadPackets();

        client.ReceivePackets();

        client.CheckForTimeOut();

        if ( client.IsDisconnected() )
            break;

        time += deltaTime;

        client.AdvanceTime( time );

        clientTransport.AdvanceTime( time );

        if ( client.ConnectionFailed() )
            break;

        platform_sleep( deltaTime );
    }
```

Conceptually, what functionality does the client perform compared to the transport? What does `client.SendPackets()`do exactly as compared to`clientTransport.WritePackets()`? 

How can user packets (type=`TEST_USER_PACKET`) be exchanged once the connection is setup?

Also, my understanding is that various user packet formats are defined in `shared.h` and a factory for these types is generated with `YOJIMBO_PACKET_FACTORY_START/STOP` macro block?

It'd be great to have some kind of HOWTO on using the library!

**Comment by gafferongames** [comment]
Hi Mike, thanks for your feedback. I agree, the library needs usage documentation badly. I've been working on some doxygen documentation for the past few weeks that should help as a reference, and after that I'd like to create some usage documentation that is much more high level, a getting started, concepts and so on. Basically pretty much what you are looking for.

Until then, here is the quick gist.

The core concepts are:

1. Client
2. Server
3. Transport
4. Connection
5. Channels
6. Packets
7. Messages

The client and server are protocol level. They create the abstraction of the server with n slots for clients to connect to, and make unreliable UDP look like a connection based protocol by implementing a connection negotiation state machine and timeouts.

The transport is the actual, on the wire sending and receiving packets. The client and server work on top of the transport, and don't actually serialize (read or write) packets, it just creates those packet objects, and adds them to a queue to be sent later, vice-versa for receive packets are read in from the network and added to a queue as packet objects, so they can be read off that queue later. This architecture is designed for adding multithreading later on.

The connection is the thing that is sending the connection packets you are seeing. This is the layer that implements acks and reliability for messages. Connections may be configured to have multiple channels. Each channel has its own send and receive queue for messages, and may have different reliability and ordering guarantees. Currently, there are reliable-ordered channels and unreliable-unordered channels.

Reliable ordered channels are good for sending control messages and RPCs, as well as large blocks of data (larger than your max packet size, they get fragmented and split up for you, and reassembled automatically on the other side).

Unreliable unordered channels are basically UDP, so messages are unreliable and may arrive out of order. Large blocks can still be sent over this channel, but in this case, there is no fragmentation support, so you have to configure your max packet size to be larger that the largest block you want to send. I'm planning on adding transport level packet fragmentation and reassembly shortly, so you have a way to send large blocks (say 2-4k) over this channel, while avoiding sending packets over the wire that are larger than MTU (~1500 bytes).

So your main choice right now is, do you extend the protocol by adding your own message types, or do you extend it by adding user packets? My recommendation is to go with messages, because that's the most fully featured path, and user messages are just meant for really low-level reasons when you really, really want your own specific packet types. You should be able to see client/server examples in test.cpp that send and receive messages, and for the user packet path look at test_client_server_user_packets example in test.cpp

Finally, the client and server are designed to be overridden to use, so take what is in shared.h and copy that to your own code and cut down, take the packet and message factories as well and cut and paste and create your own set of messages for stuff you want, and packets (if you want that), the implementation in shared.h is mostly there to support the tests in test.cpp and the testbed programs I use like client/server/secure_client/secure_server, but you should be able to quickly cut them down to just what you need.

Transport and the other classes you can use as is, there is no need to subclass them. FWIW, the macros around the packet factory and message factory, are wrapping up a subclass of the PacketFactory and MessageFactory classes, to create one customized to the set of packets and messages you define.

cheers

- Glenn

**Comment by gafferongames** [comment]
Also, to answer your question specifically about the difference between the client/server and transport behavior, the Server::SendPackets, Client::SendPackets is really just generating the packets and queuing them on the transport. They don't actually get written to the network until Transport::WritePackets is called.

Similarly, Server::ReceivePackets, Client::ReceivePackets don't actually read packets from the network, they just pop packets off the transport receive queue, so you want to call Transport::ReadPackets (which does), before that, to make sure you have the most recent packets in the receive queue first.

cheers

- Glenn

**Comment by mikesun** [comment]
Thanks Glenn, your response is super helpful and clarifies quite a bit for me.

I took a closer look at `test.cpp::test_client_server_messages()`. The call to the `PumpConnectionUpdate()` is basically equivalent to this block in `client.cpp` right?:

```
        client.SendPackets();

        clientTransport.WritePackets();

        clientTransport.ReadPackets();

        client.ReceivePackets();

        client.CheckForTimeOut();

        if ( client.IsDisconnected() )
            break;

        time += deltaTime;

        client.AdvanceTime( time );

        clientTransport.AdvanceTime( time );
```

It seems that the client/server implementation is responsible for regularly calling `SendPackets(),WritePackets()` and `ReadPackets(),ReceivePackets`; basically generating and reading the "wave" of packets on which messages/acks are piggybacked, yes?

Given that the implementation is currently not multi-threaded, does that mean that any sending/receiving message functions must be called from within the same loop that generates the packet wave (as opposed to another thread apart from the one running the packet wave loop)?

Lastly, it seems network IO performed by WritePackets()/ReadPackets() is non-blocking, but not async (no use of libevent or select/epoll/kqueue). Is that choice made because it's assumed socket reads/writes will occur continuously?

Thanks again for the quick and informative response!

**Comment by gafferongames** [comment]
Yes, the library is designed around common patterns used in dedicated server-based games like FPS. These games are simulation, not I/O bound, and completely stateful, so a async IO doesn't make a lot of sense.

The basic pattern is:

start of game frame on server
read packets from network
process packets from client (eg. receive input packets), and process them.
step forward players according to inputs
step forward world simulation one tick
delta encode and contruct stateful packets to send out to clients with state of current world at time t
flush packets to network
(sleep until next, 10HZ, 20HZ, 30HZ or 60HZ server frame)

cheers

**Comment by gafferongames** [comment]
> My understanding is that the client/server implementation is responsible for regularly calling 
> SendPackets(),WritePackets() and ReadPackets(),ReceivePackets --- generating and reading the 
> "wave" of packets on which messages/acks are piggybacked. Is that correct?

Yes. You have to pump this, the basic loop is going to look pretty similar for all examples. I mentioned that test because if you follow the code, you'll find the bits of code that send and receive user packets.

cheers

**Comment by gafferongames** [comment]
> Given that the implementation is currently not multi-threaded, does that mean that any send/receive
> message functions must be called from within the same loop that generates the packet wave (as 
> opposed to another thread apart from the one running the packet wave loop)?

Yes.

**Comment by gafferongames** [comment]
> Also, it seems network IO performed by WritePackets()/ReadPackets() is non-blocking, but not 
> async (no use of something like libevent or select/epoll/kqueue). Is that choice made because: 
> it's assumed socket reads will occur continuously, essentially polling at the application-level;
>  and that transport level acks provide for guarantee of transmission?

It's because the application this library is designed for is frame based, so it's typical for you to want to pump all packets received since last frame, process them, *do stuff*, send packets out, sleep until next frame.

Contrast with an async loop where its like, setup a bunch of worker threads, sleep until packets come in, holy shit a packet has come in do some work on some thread, maybe reply with a packet. Rinse and repeat for each packet that comes in, hopefully distributing all work across a bunch of cores etc.

So basically, frame based vs. packet based is the reason.

cheers

- Glenn

**Comment by gafferongames** [comment]
I think you have the information you need, so I'm closing this issue. Thanks for asking for the information you needed and feel free to keep asking questions in here if you have more.

cheers

- Glenn
